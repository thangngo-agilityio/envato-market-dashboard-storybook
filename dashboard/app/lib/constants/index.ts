export * from './routers';
export * from './time';
export * from './images';
export * from './form';
export * from './messages';
export * from './regex';
export * from './texts';
export * from './themes';
export * from './api';
export * from './images';
export * from './menu';
export * from './sidebar';
export * from './charts';
export * from './sidebar';
export * from './common';
export * from './options';
export * from './transaction';
export * from './store';
export * from './status';
export * from './pagination';
export * from './columns';
export * from './faq';
export * from './avatar';
export * from './sizes';
export * from './setting';
export * from './termAndCondition';
export * from './notification';
export * from './authentication';
export * from './discount';
export * from './metadata';
export * from './og-properties';
export * from './chat';
export * from './product';
