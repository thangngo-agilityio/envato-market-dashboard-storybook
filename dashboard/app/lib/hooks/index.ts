export * from './useForm';
export * from './useAuth';
export * from './useStatistical';
export * from './useTransactions';
export * from './useDebounce';
export * from './useSearch';
export * from './usePagination';
export * from './useStore';
export * from './useUser';
export * from './usePinCode';
export * from './useNotification';
export * from './useWallet';
export * from './useMoney';
export * from './useChat';
export * from './useImageUploader';
export * from './useProducts';
export * from './useRecentActivities';
