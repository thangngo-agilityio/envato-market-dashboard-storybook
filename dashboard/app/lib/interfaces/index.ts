export * from './mock';
export * from './user';
export * from './sort';
export * from './icons';
export * from './spending';
export * from './status';
export * from './transaction';
export * from './table';
export * from './config';
export * from './pagination';
export * from './notification';
export * from './chat';
export * from './issues';
export * from './form';
export * from './wallet';
export * from './money';
export * from './products';
export * from './recentActivities';
