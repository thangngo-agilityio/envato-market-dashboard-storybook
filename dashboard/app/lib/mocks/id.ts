export const CHAT_ID =
  'eIrqNbRtJkawzyFxgHjRLGq2LGn191xfYluhBNQtxkoMoJGrkQdrOwv1';
export const SENDER_ID = 'eIrqNbRtJkawzyFxgHjRLGq2LGn1';

export const USER_ID = '91xfYluhBNQtxkoMoJGrkQdrOwv1';
