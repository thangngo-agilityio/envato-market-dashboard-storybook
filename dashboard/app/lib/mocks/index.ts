export * from './spending';
export * from './table';
export * from './transaction';
export * from './product';
export * from './user';
export * from './time';
export * from './statistic';
export * from './chat';
export * from './issues';
export * from './wallet';
export * from './supports';
export * from './recentActivities';
