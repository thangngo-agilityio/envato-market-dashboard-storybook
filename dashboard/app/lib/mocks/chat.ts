// Constants
import { AVATAR_POSITION, IMAGES } from '@/lib/constants';
import { MESSAGE_TIME } from '.';

// TODO: Update later
export const USER_CHATS = [
  {
    uid: 'user1',
    isSend: false,
    content: 'Hi, I need more information',
    avatarPosition: AVATAR_POSITION.BEFORE,
    avatar: IMAGES.CHAT_USER_AVATAR.url,
    localeTime: MESSAGE_TIME + 1000,
  },

  {
    uid: 'user2',
    isSend: false,
    content: 'Hello, I want to know more about your product',
    avatarPosition: AVATAR_POSITION.BEFORE,
    avatar: IMAGES.CHAT_USER_AVATAR.url,
    localeTime: MESSAGE_TIME + 3000,
  },

  {
    uid: 'user3',
    isSend: false,
    content: 'Hello, I want to know more about your product',
    avatarPosition: AVATAR_POSITION.BEFORE,
    avatar: IMAGES.CHAT_USER_AVATAR.url,
    localeTime: MESSAGE_TIME + 6000,
  },

  {
    uid: 'admin1',
    isSend: true,
    content: 'Hi, i am here for help',
    avatarPosition: AVATAR_POSITION.AFTER,
    avatar: IMAGES.CHAT_USER_AVATAR.url,
    localeTime: MESSAGE_TIME + 2000,
  },

  {
    uid: 'admin2',
    isSend: true,
    content: 'Sure, I can help you with that',
    avatarPosition: AVATAR_POSITION.BEFORE,
    avatar: IMAGES.CHAT_USER_AVATAR.url,
    localeTime: MESSAGE_TIME + 4000,
  },
  {
    uid: 'admin3',
    isSend: true,
    content: 'Sure, I can help you with that',
    avatarPosition: AVATAR_POSITION.BEFORE,
    avatar: IMAGES.CHAT_USER_AVATAR.url,
    localeTime: MESSAGE_TIME + 7000,
  },
];

export const MEMBER_CHATS = [
  { uid: 'admin', name: 'Admin' },
  { uid: 'user1', name: 'User 1' },
];

export const CONVERSATION_PROPS = {
  messages: [
    {
      date: {
        nanoseconds: 1,
        seconds: 11,
      },
      id: '1',
      senderId: '2',
      text: 'Hello',
    },
  ],
};
