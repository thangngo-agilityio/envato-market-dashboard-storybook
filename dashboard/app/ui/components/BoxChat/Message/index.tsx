import { Box, Flex, Spacer, Text } from '@chakra-ui/react';
import { useEffect, useState } from 'react';
import Image from 'next/image';

// Constants
import { IMAGES } from '@/lib/constants';

// Utils
import { generatePlaceholder } from '@/lib/utils';

interface MessageProps {
  content?: string;
  avatarUser?: string;
  avatarAdmin?: string;
  localeTime?: string;
  senderId: string;
  superAdminId: string;
}

const Message = ({
  senderId,
  content = '',
  avatarUser = IMAGES.CHAT_USER_AVATAR.url,
  avatarAdmin = IMAGES.CHAT_USER_AVATAR.url,
  localeTime = new Date().toLocaleTimeString([], {
    hour: 'numeric',
    minute: '2-digit',
  }),
  superAdminId,
}: MessageProps) => {
  const [adminId, setAdminId] = useState<string>('');

  useEffect(() => {
    const fetchData = async () => {
      setAdminId(superAdminId || '');
    };

    fetchData();
  }, []);

  const isAdmin = senderId === adminId;

  return (
    <Flex width="100%" justifyContent={!isAdmin ? 'flex-end' : 'flex-start'}>
      {isAdmin && (
        <Box pos="relative" w={9} h={9}>
          <Image
            src={avatarAdmin}
            alt={avatarAdmin}
            fill
            sizes="100vw"
            placeholder="blur"
            blurDataURL={generatePlaceholder(36, 36)}
            style={{
              borderRadius: '50%',
              marginRight: '8px',
            }}
          />
        </Box>
      )}

      <Flex
        align="flex-end"
        direction={!isAdmin ? 'row-reverse' : 'row'}
        mb="30px"
        alignItems="center"
      >
        <Box
          data-testid="image-container"
          bg={!isAdmin ? 'primary.300' : 'background.section.messageUser'}
          p={3}
          ml={2}
          borderRadius={8}
          color="text.primary"
          dangerouslySetInnerHTML={{ __html: content ?? '' }}
        />
        <Spacer />

        <Text
          ml={3}
          mr={3}
          fontSize="xs"
          color="text.textTime"
          fontWeight="medium"
          minW="max-content"
        >
          {localeTime}
        </Text>
      </Flex>

      {!isAdmin && (
        <Box pos="relative" w={9} h={9}>
          <Image
            src={avatarUser}
            alt={avatarUser}
            fill
            sizes="100vw"
            placeholder="blur"
            blurDataURL={generatePlaceholder(36, 36)}
            style={{
              borderRadius: '50%',
              marginLeft: '8px',
            }}
          />
        </Box>
      )}
    </Flex>
  );
};
export default Message;
