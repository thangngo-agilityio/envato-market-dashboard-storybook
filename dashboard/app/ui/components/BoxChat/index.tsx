'use client';

import { memo, useCallback, useEffect, useRef, useState } from 'react';
import { Box, Heading, Flex } from '@chakra-ui/react';

// Components
import Message from './Message';

// Components
import { InputSendMessages } from '@/ui/components';

// Interface
import { TMessages } from '@/lib/interfaces';

// Stores
import { authStore } from '@/lib/stores';

// Hooks
import { getInfoRoomChat, useSubscribeToChat } from '@/lib/hooks';

// Firebase
import { doc, getDoc, setDoc } from 'firebase/firestore';
import { db } from '@/lib/utils';
import { AUTHENTICATION_ROLE, FIREBASE_CHAT } from '@/lib/constants';

const initialUserChat = {
  roomChatId: '',
  userId: '',
  adminId: '',
  avatarUrl: '',
  avatarAdminUrl: '',
  displayName: '',
};

const BoxChatComponent = () => {
  const [messages, setMessages] = useState<TMessages[]>([]);
  const user = authStore((state) => state.user);
  const hasPermission = user?.role === AUTHENTICATION_ROLE.MEMBER;
  const [userChat, setUserChat] = useState(initialUserChat);
  const boxRef = useRef<HTMLDivElement | null>(null);

  const fetchData = async () => {
    const usersData = await getInfoRoomChat(user);
    // Get user data

    // Check if usersData is undefined before accessing its properties
    if (usersData) {
      const res = await getDoc(
        doc(db, FIREBASE_CHAT.CHATS, usersData.roomChatId),
      );

      setUserChat(usersData);
      res.exists() ? setMessages(res.data().messages) : await createChatRoom();
    }
  };

  const createChatRoom = useCallback(async () => {
    // Get user data
    const usersData = await getInfoRoomChat(user);

    if (usersData) {
      await setDoc(doc(db, FIREBASE_CHAT.CHATS, usersData.roomChatId), {
        messages: [],
      });
    }
  }, [user]);

  useEffect(() => {
    fetchData();
  }, []);

  useSubscribeToChat(userChat.roomChatId, setMessages, boxRef);

  return (
    hasPermission && (
      <Box w="full" bg="background.body.quaternary" borderRadius="lg">
        <Flex
          direction="row"
          justifyContent="space-between"
          alignItems="center"
          justify="center"
          borderBottom="1px solid"
          borderBottomColor="border.tertiary"
          padding="24px 26px"
        >
          <Heading
            as="h3"
            fontWeight="semibold"
            color="text.primary"
            fontSize="2xl"
            textTransform="capitalize"
          >
            team chat
          </Heading>
        </Flex>

        <Box padding={{ base: '24px 20px', lg: '45px 35px' }}>
          <Box
            ref={boxRef}
            overflowX="auto"
            overflowY="scroll"
            css={{
              '&::-webkit-scrollbar': {
                width: 2,
              },
              '&::-webkit-scrollbar-track': {
                width: 2,
              },
              '&::-webkit-scrollbar-thumb': {
                background: 'gray',
                borderRadius: '24px',
              },
            }}
            maxHeight={361}
            padding={5}
          >
            {messages.map((message) => (
              <Message
                content={message.text}
                key={message.date.seconds}
                senderId={message.senderId}
                avatarAdmin={userChat.avatarAdminUrl}
                avatarUser={userChat.avatarUrl}
                superAdminId={userChat.adminId}
              />
            ))}
          </Box>
          <InputSendMessages boxRef={boxRef} />
        </Box>
      </Box>
    )
  );
};

const BoxChat = memo(BoxChatComponent);

export default BoxChat;
