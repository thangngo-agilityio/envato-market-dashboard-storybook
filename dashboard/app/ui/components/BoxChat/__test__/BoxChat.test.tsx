// Components
import BoxChat from '@/ui/components/BoxChat';

// Stores
import { authStore } from '@/lib/stores';

// Mocks
import { MOCK_USER_DETAIL, USER_DATA } from '@/lib/mocks';

const getInfoRoomChatMock = jest.fn();

jest.mock('@/lib/hooks', () => ({
  ...jest.requireActual('@/lib/hooks'),
  getInfoRoomChat: () => getInfoRoomChatMock,
}));

const roomChatId = 'GmCJFXqXubfAPdKs56C4Sq7DisY2lBNWnoNBQGZ260KgnF9NxQCPlqf1';

describe('BoxChatComponent', () => {
  beforeEach(() => {
    authStore.setState({
      user: MOCK_USER_DETAIL,
    });
  });
  test('BoxChat component renders user have roomChatId', () => {
    getInfoRoomChatMock.mockReturnValue({ ...USER_DATA, roomChatId });
    const { container } = render(<BoxChat />);
    expect(container).toMatchSnapshot();
  });

  test('BoxChat component renders user without roomChatId', () => {
    getInfoRoomChatMock.mockReturnValue(USER_DATA);
    const { container } = render(<BoxChat />);
    expect(container).toMatchSnapshot();
  });
});
