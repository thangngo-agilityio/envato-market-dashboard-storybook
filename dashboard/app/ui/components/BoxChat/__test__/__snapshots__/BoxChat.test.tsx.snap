// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BoxChatComponent BoxChat component renders user have roomChatId 1`] = `
<div>
  <div
    class="css-1pphdyr"
  >
    <div
      class="css-z7702n"
    >
      <h3
        class="chakra-heading css-1rl4kbt"
      >
        team chat
      </h3>
    </div>
    <div
      class="css-j9vbtn"
    >
      <div
        class="css-3htvxw"
      />
      <form
        class="chakra-stack css-tl3ftk"
      >
        <div
          class="css-1jif775"
        >
          <div
            class="chakra-form-control css-1kxonj9"
            role="group"
          >
            <div
              class="css-1yae3jf"
            >
              <div
                class="chakra-form-control css-1kxonj9"
                role="group"
              >
                <div
                  class="chakra-input__group css-1091msq"
                  data-group="true"
                >
                  <input
                    class="chakra-input css-1s8wmp9"
                    id="field-:r2:"
                    name="text"
                    placeholder="Type your message..."
                    type="text"
                    value=""
                  />
                </div>
              </div>
            </div>
          </div>
          <p
            class="chakra-text css-rj1t1b"
          />
          <button
            class="chakra-button css-bg2fy7"
            data-testid="btn-send"
            disabled=""
            type="submit"
          >
            <svg
              fill="none"
              height="18"
              viewBox="0 0 20 18"
              width="20"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M17.3894 0H2.61094C0.339326 0 -0.844596 2.63548 0.696196 4.26234L3.78568 7.52441C4.23 7.99355 4.47673 8.60858 4.47673 9.24704V15.4553C4.47673 17.8735 7.61615 18.9233 9.13941 17.0145L19.4463 4.09894C20.7775 2.43071 19.5578 0 17.3894 0Z"
                fill="#22C55E"
              />
            </svg>
          </button>
        </div>
      </form>
    </div>
  </div>
  <span
    hidden=""
    id="__chakra_env"
  />
</div>
`;

exports[`BoxChatComponent BoxChat component renders user without roomChatId 1`] = `
<div>
  <div
    class="css-1pphdyr"
  >
    <div
      class="css-z7702n"
    >
      <h3
        class="chakra-heading css-1rl4kbt"
      >
        team chat
      </h3>
    </div>
    <div
      class="css-j9vbtn"
    >
      <div
        class="css-3htvxw"
      />
      <form
        class="chakra-stack css-tl3ftk"
      >
        <div
          class="css-1jif775"
        >
          <div
            class="chakra-form-control css-1kxonj9"
            role="group"
          >
            <div
              class="css-1yae3jf"
            >
              <div
                class="chakra-form-control css-1kxonj9"
                role="group"
              >
                <div
                  class="chakra-input__group css-1091msq"
                  data-group="true"
                >
                  <input
                    class="chakra-input css-1s8wmp9"
                    id="field-:r5:"
                    name="text"
                    placeholder="Type your message..."
                    type="text"
                    value=""
                  />
                </div>
              </div>
            </div>
          </div>
          <p
            class="chakra-text css-rj1t1b"
          />
          <button
            class="chakra-button css-bg2fy7"
            data-testid="btn-send"
            disabled=""
            type="submit"
          >
            <svg
              fill="none"
              height="18"
              viewBox="0 0 20 18"
              width="20"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M17.3894 0H2.61094C0.339326 0 -0.844596 2.63548 0.696196 4.26234L3.78568 7.52441C4.23 7.99355 4.47673 8.60858 4.47673 9.24704V15.4553C4.47673 17.8735 7.61615 18.9233 9.13941 17.0145L19.4463 4.09894C20.7775 2.43071 19.5578 0 17.3894 0Z"
                fill="#22C55E"
              />
            </svg>
          </button>
        </div>
      </form>
    </div>
  </div>
  <span
    hidden=""
    id="__chakra_env"
  />
</div>
`;
